marlowe-playground-client/src/Marlowe/Linter.purs

```diff
@@ -138,198 +147,427 @@ termToRange a { row, column } =
+  _ <- traverse (lintCase env) cases
+  _ <- CMS.modify (over _holes (insertHole hole))
+  _ <- lintContract env cont
+  pure unit
+
+lintContract env (Term (When cases timeoutTerm@(Term timeout pos) cont) _) = do
   let
-    addWarnings = if Set.member valueId (view _letBindings state) then identity else Set.insert (UninitializedUse (termToRange valueId pos))
-  in
-    state
-      # over _holes (getHoles valueId)
-      # over _warnings addWarnings
+    timeoutNotIncreasing = if timeout > (view _maxTimeout env) then mempty else Set.singleton (TimeoutNotIncreasing (termToRange timeout pos))
 
-lintValue state (Term (UseValue hole) _) = over _holes (insertHole hole) state
+    newEnv = (over _maxTimeout (max timeout)) env
+  _ <- traverse (lintCase newEnv) cases
+  _ <- CMS.modify (over _holes (insertHole timeoutTerm))
+  _ <- CMS.modify (over _warnings (Set.union timeoutNotIncreasing))
+  _ <- lintContract newEnv cont
+  pure unit
 
-lintValue state hole@(Hole _ _ _) = over _holes (insertHole hole) state
-
-maybeInsert :: forall a. Ord a => Maybe a -> Set a -> Set a
-maybeInsert Nothing xs = xs
-
-maybeInsert (Just x) xs = Set.insert x xs
-
-collectFromTuples :: forall a b. Array (a /\ b) -> Array a /\ Array b
-collectFromTuples = foldMap (\(a /\ b) -> [ a ] /\ [ b ])
-
-lintCase :: State -> Term Case -> State /\ Maybe (Term Contract)
-lintCase state (Term (Case action contract) _) =
+lintContract env (Term (Let valueIdTerm@(Term valueId pos) value cont) _) = do
   let
-    newState =
-      state
-        # over _warnings (maybeInsert (negativeDeposit action))
-  in
-    lintAction newState action /\ Just contract
+    shadowedLet = if Set.member valueId (view _letBindings env) then Set.singleton (ShadowedLet (termToRange valueId pos)) else mempty
 
-lintCase state hole@(Hole _ _ _) = over _holes (insertHole hole) state /\ Nothing
+    newEnv = over _letBindings (Set.insert valueId) env
+  _ <- CMS.modify (over _warnings (Set.union shadowedLet))
+  sa <- lintValue env value
+  markSimplification constToVal SimplifiableValue value sa
+  lintContract newEnv cont
 
-lintAction :: State -> Term Action -> State
-lintAction state (Term (Deposit acc party token value) _) =
+lintContract env (Term (Let valueIdTerm@(Hole _ _ _) value cont) _) = do
   let
-    gatherHoles = getHoles acc <> getHoles party <> getHoles token
-
-    newState = over _holes (gatherHoles) state
+    gatherHoles = getHoles valueIdTerm
+  _ <- CMS.modify (over _holes gatherHoles)
+  sa <- lintValue env value
+  markSimplification constToVal SimplifiableValue value sa
+  lintContract env cont
+
+lintContract env hole@(Hole _ _ _) = do
+  _ <- CMS.modify (over _holes (insertHole hole))
+  pure unit
+
+lintObservation :: LintEnv -> Term Observation -> CMS.State State (TemporarySimplification Boolean Observation)
+lintObservation env t@(Term (AndObs a b) pos) = do
+  sa <- lintObservation env a
+  sb <- lintObservation env b
+  case sa /\ sb of
```

Comment:
I _think_ in purescript you can do `case sa sb of`

Reply:
I cannot get it to work :/

Reply:
ah https://github.com/purescript/documentation/blob/master/language/Syntax.md#case-expressions shows that it's `case a, b of`
It doesn't look much simpler than `/\` however it is syntax rather than a function, not sure if this is better or not but it's there for you if you want it.